x/ccv/provider/keeper/consumer_lifecycle.go

```diff
@@ -0,0 +1,670 @@
+	if err != nil {
+		return gen, nil, errorsmod.Wrapf(sdkerrors.ErrInvalidHeight, "error %s calculating trusting_period for: %s", err, height)
+	}
+	clientState.TrustingPeriod = trustPeriod
+	clientState.UnbondingPeriod = providerUnbondingPeriod
+
+	consState, err := k.clientKeeper.GetSelfConsensusState(ctx, height)
+	if err != nil {
+		return gen, nil, errorsmod.Wrapf(clienttypes.ErrConsensusStateNotFound, "error %s getting self consensus state for: %s", err, height)
+	}
+
+	// get the bonded validators from the staking module
+	bondedValidators, err := k.GetLastBondedValidators(ctx)
+	if err != nil {
+		return gen, nil, errorsmod.Wrapf(stakingtypes.ErrNoValidatorFound, "error getting last bonded validators: %s", err)
+	}
+
+	minPower := int64(0)
+	if powerShapingParameters.Top_N > 0 {
+		// get the consensus active validators
+		// we do not want to base the power calculation for the top N
+		// on inactive validators, too, since the top N will be a percentage of the active set power
+		// otherwise, it could be that inactive validators are forced to validate
+		activeValidators, err := k.GetLastProviderConsensusActiveValidators(ctx)
+		if err != nil {
+			return gen, nil, errorsmod.Wrapf(stakingtypes.ErrNoValidatorFound, "error getting last active bonded validators: %s", err)
+		}
+
+		// in a Top-N chain, we automatically opt in all validators that belong to the top N
+		minPower, err = k.ComputeMinPowerInTopN(ctx, activeValidators, powerShapingParameters.Top_N)
+		if err != nil {
+			return gen, nil, err
+		}
+		// log the minimum power in top N
+		k.Logger(ctx).Info("minimum power in top N at consumer genesis",
+			"consumerId", consumerId,
+			"minPower", minPower,
+		)
+		k.OptInTopNValidators(ctx, consumerId, activeValidators, minPower)
+		k.SetMinimumPowerInTopN(ctx, consumerId, minPower)
+	}
+
+	// need to use the bondedValidators, not activeValidators, here since the chain might be opt-in and allow inactive vals
+	nextValidators := k.ComputeNextValidators(ctx, consumerId, bondedValidators, powerShapingParameters, minPower)
+	k.SetConsumerValSet(ctx, consumerId, nextValidators)
+
+	// get the initial updates with the latest set consumer public keys
+	initialUpdatesWithConsumerKeys := DiffValidators([]types.ConsensusValidator{}, nextValidators)
+
+	// Get a hash of the consumer validator set from the update with applied consumer assigned keys
+	updatesAsValSet, err := tmtypes.PB2TM.ValidatorUpdates(initialUpdatesWithConsumerKeys)
+	if err != nil {
+		return gen, nil, fmt.Errorf("unable to create validator set from updates computed from key assignment in MakeConsumerGenesis: %s", err)
+	}
+	hash := tmtypes.NewValidatorSet(updatesAsValSet).Hash()
+
+	consumerGenesisParams := ccv.NewParams(
+		true,
+		initializationRecord.BlocksPerDistributionTransmission,
+		initializationRecord.DistributionTransmissionChannel,
+		"", // providerFeePoolAddrStr,
```

Comment:
I feel like this needs a bit more of a comment here, looks a bit sketchy and like someone just forgot to uncomment this without any extra comment